decorators.ts

```diff
@@ -3,35 +3,69 @@
 
 import Promise = require("bluebird");
 import fiberBootstrap = require("./fiber-bootstrap");
+import * as assert from "assert";
 
-export function exported(moduleName: string) {
-	return (target: Object, propertyKey: string, descriptor: TypedPropertyDescriptor<any>) => {
+export function exportedPromise(moduleName: string): any {
+	return (target: Object, propertyKey: string, descriptor: TypedPropertyDescriptor<any>): TypedPropertyDescriptor<any> => {
 		$injector.publicApi.__modules__[moduleName] = $injector.publicApi.__modules__[moduleName] || {};
-		$injector.publicApi.__modules__[moduleName][propertyKey] = (...args: any[]): any => {
-			return new Promise(function(onFulfilled : Function, onRejected: Function) {
-					let originalModule = $injector.resolve(moduleName);
-					let originalMethod: Function = originalModule[propertyKey];
-					let result: any;
-					try {
-						result = originalMethod.apply(originalModule, args);
-					} catch(err) {
-						onRejected(err);
-						return;
-					}
-
-					if(result && typeof result.wait === "function") {
-						fiberBootstrap.run(function () {
-							try {
-								let realResult = result.wait();
-								onFulfilled(realResult);
-							} catch(err) {
-								onRejected(err);
-							}
-						});
-					} else {
-						onFulfilled(result);
-					}
+		$injector.publicApi.__modules__[moduleName][propertyKey] = (...args: any[]): Promise<any>[] | Promise<any>=> {
+			let originalModule = $injector.resolve(moduleName);
+			let originalMethod: Function = originalModule[propertyKey];
+			let result: any;
+			try {
+				result = originalMethod.apply(originalModule, args);
+			} catch(err) {
+				let promise = new Promise(function(onFulfilled : Function, onRejected: Function) {
+					onRejected(err);
 				});
+
+				// TODO: Check if we should return [promise] or promise.
+				return promise;
+			}
+
+			let types = _(result)
+						.groupBy(f => typeof f)
+						.keys()
+						.value();
+
+			// Check if method returns IFuture<T>[]. In this case we will return Promise<T>[]
+			if(_.isArray(result) && types.length === 1 && typeof(_.first<any>(result).wait) === "function") {
```

Comment:
The check whether a value is a future happens on multiple places. Is it profitable to extract it in a dedicated function?